File: CPlusPlus/1674_minimum_moves.cpp
class Solution {
public:
    int minMoves(vector<int>& nums, int limit) {
        int N = 2 * limit + 2;
        int M = nums.size();
        vector<int> events(N, 0);

        vector<pair<int, int>> pairs;
        for (int i = 0; i < nums.size() / 2; i++) {
            int x = nums[i], y = nums[M - i - 1];
            if (x > y) swap(x, y);
            pairs.push_back({x, y});
        }
        for (auto& pair : pairs) {
            int x = pair.first, y = pair.second;
            events[0] += 2;
            events[N - 1] -= 2;

            events[x + 1] -= 1;
            events[y + limit + 1] += 1;

            events[x + y] -= 1;
            events[x + y + 1] += 1;
        }
        
        for (int i = 1; i < N; i++) {
            events[i] += events[i - 1];
        }
        int result = INT_MAX;
        for (int i = 0; i < N - 1; i++){
            result = min(result, events[i]);
        }
        return result;
    }
};

File: Python/1674_minimum_moves.py
class Solution:
    def minMoves(self, nums: List[int], limit: int) -> int:
        events = [0] * (2 * limit + 2)
        
        pairs = []
        for i in range(len(nums) >> 1):
            x, y = nums[i], nums[~i]
            if x > y:
                x, y = y, x
            pairs.append((x, y))
            
        for x, y in pairs:
            # Add 2 to everything
            events[0] += 2
            events[-1] -= 2

            # Add 1 to events[x + 1 .. y + limit]
            events[x + 1] -= 1
            events[y + limit + 1] += 1
            
            # Subtract 1 from events[x + y]
            events[x + y] -= 1
            
            # Add 1 from events[x + y + 1]
            events[x + y + 1] += 1
        
        for i in range(1, len(events)):
            events[i] += events[i - 1]
            
        return min(events[:-1])
